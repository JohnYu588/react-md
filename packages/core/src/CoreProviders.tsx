"use client";

import { type ReactElement, type ReactNode } from "react";

import { SsrProvider } from "./SsrProvider.js";
import { UserInteractionModeProvider } from "./interaction/UserInteractionModeProvider.js";
import { AppSizeProvider } from "./media-queries/AppSizeProvider.js";
import {
  type AppSize,
  type AppSizeQueries,
  DEFAULT_APP_SIZE,
  DEFAULT_APP_SIZE_QUERIES,
} from "./media-queries/appSize.js";
/* eslint-disable @typescript-eslint/no-unused-vars */
import { type MenuConfigurationProvider } from "./menu/MenuConfigurationProvider.js";
import {
  type PortalContainer,
  PortalContainerProvider,
} from "./portal/PortalContainerProvider.js";
import { type Snackbar } from "./snackbar/Snackbar.js";
import { type LocalStorageColorSchemeProvider } from "./theme/LocalStorageColorSchemeProvider.js";
import { type ThemeProvider } from "./theme/ThemeProvider.js";
import { type useColorSchemeProvider } from "./theme/useColorSchemeProvider.js";
import { type TooltipHoverModeProvider } from "./tooltip/TooltipHoverModeProvider.js";
import { type WritingDirectionProvider } from "./typography/WritingDirectionProvider.js";

/* eslint-enable @typescript-eslint/no-unused-vars */

/** @since 6.0.0 */
export interface ReactMDCoreConfiguration {
  /**
   * Set this to `true` if you are using SSR. This will update some hook
   * behavior so that you will not have mismatched initial state between server
   * and client and correctly rehydrate after initial render.
   *
   * @defaultValue `false`
   */
  ssr?: boolean;

  /**
   * @defaultValue `DEFAULT_APP_SIZE`
   * @see {@link DEFAULT_APP_SIZE}
   */
  ssrAppSize?: Readonly<AppSize>;

  /**
   * @defaultValue `DEFAULT_APP_SIZE_QUERIES`
   * @see {@link DEFAULT_APP_SIZE_QUERIES}
   */
  appSizeQueries?: Readonly<AppSizeQueries>;

  /** @see {@link PortalContainerProvider} */
  portalContainer?: PortalContainer;
}

/** @since 6.0.0 */
export interface CoreProvidersProps extends ReactMDCoreConfiguration {
  children: ReactNode;
}

/**
 * **Client Component**
 *
 * This is a convenience wrapper for initializing providers for react-md. This
 * includes configuration for:
 *
 * - {@link SsrProvider}
 * - {@link PortalContainerProvider}
 * - {@link AppSizeProvider}
 * - {@link UserInteractionModeProvider}
 *
 * This does not include every provider within `react-md` to help decrease the
 * bundle size for unused features. Here are some other providers that might be
 * useful to initialize near the root of your app:
 *
 * - {@link MenuConfigurationProvider} - If you want to render menus as sheets
 *   on phones or other configuration.
 * - {@link LocalStorageColorSchemeProvider}/{@link useColorSchemeProvider} - If
 *   you want to allow the user to select a light, dark, or system theme
 * - {@link WritingDirectionProvider} - If you need dynamic support for ltr and
 *   rtl languages
 * - {@link ThemeProvider} - If you allow the user to dynamically configure the
 *   `react-md` theme
 * - {@link TooltipHoverModeProvider} - If you want tooltips to appear
 *   immediately for a time after another tooltip has become visible. You can
 *   also configure the global visible delay.
 * - {@link Snackbar} - This isn't a provider but this is how you can display
 *   alerts in your app
 *
 * @see {@link https://next.react-md.dev/components/core-providers | CoreProviders Demos}
 * @since 6.0.0
 */
export function CoreProviders(props: CoreProvidersProps): ReactElement {
  const {
    ssr = false,
    ssrAppSize = DEFAULT_APP_SIZE,
    appSizeQueries = DEFAULT_APP_SIZE_QUERIES,
    portalContainer,
    children,
  } = props;
  return (
    <SsrProvider ssr={ssr}>
      <PortalContainerProvider container={portalContainer}>
        <UserInteractionModeProvider>
          <AppSizeProvider {...appSizeQueries} ssrSize={ssrAppSize}>
            {children}
          </AppSizeProvider>
        </UserInteractionModeProvider>
      </PortalContainerProvider>
    </SsrProvider>
  );
}
