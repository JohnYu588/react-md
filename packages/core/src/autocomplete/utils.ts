import { type RefObject } from "react";
import {
  type EditableHTMLElement,
  triggerManualChangeEvent,
} from "../form/utils.js";
import { type NonNullMutableRef, type UseStateInitializer } from "../types.js";
import { noopAutocompleteFilter } from "./defaults.js";
import {
  type AutocompleteFilterOptions,
  type AutocompleteGetOptionLabel,
  type AutocompleteOption,
} from "./types.js";

/**
 * @since 6.0.0
 * @internal
 */
export interface GetDefaultValueOptions<Option extends AutocompleteOption> {
  query: string | undefined;
  multiselect?: boolean;
  defaultQuery: UseStateInitializer<string> | undefined;
  defaultValue:
    | UseStateInitializer<Option | null | readonly Option[]>
    | undefined;
  options: readonly Option[];
  getOptionLabel: AutocompleteGetOptionLabel<Option>;
  filter: (options: AutocompleteFilterOptions<Option>) => readonly Option[];
}

/**
 * @since 6.0.0
 * @internal
 */
export function getDefaultValue<Option extends AutocompleteOption>(
  options: GetDefaultValueOptions<Option>
): UseStateInitializer<Option | null | readonly Option[]> {
  const {
    query,
    filter,
    multiselect,
    defaultQuery,
    defaultValue,
    options: values,
    getOptionLabel,
  } = options;

  if (typeof defaultValue !== "undefined") {
    return defaultValue;
  }

  // do not support determining the default value for a multiselect
  // autocomplete if only a query was provided.
  if (multiselect) {
    return [];
  }

  // do not support a default value out of the box for this case
  if (filter === noopAutocompleteFilter) {
    return null;
  }

  return () => {
    let q = "";
    if (query) {
      q = query;
    } else if (defaultQuery) {
      q = defaultQuery instanceof Function ? defaultQuery() : defaultQuery;
    }

    if (!q) {
      return null;
    }

    const filtered = filter({
      list: values,
      query: q,
      extractor: getOptionLabel,
    });

    return filtered[0] ?? null;
  };
}

/**
 * @since 6.0.0
 * @internal
 */
interface GetDefaultQueryOptions<Option extends AutocompleteOption> {
  value: Option | null | readonly Option[];
  getOptionLabel: AutocompleteGetOptionLabel<Option>;
  defaultQuery?: UseStateInitializer<string>;
}

/**
 * @since 6.0.0
 * @internal
 */
export function getDefaultQuery<Option extends AutocompleteOption>(
  options: GetDefaultQueryOptions<Option>
): UseStateInitializer<string> {
  const { value, getOptionLabel, defaultQuery } = options;

  if (defaultQuery) {
    return defaultQuery;
  }

  return () => {
    if (typeof value === "string") {
      return value;
    }

    if (value && typeof value === "object" && !("length" in value)) {
      return getOptionLabel(value);
    }

    return "";
  };
}

/**
 * @since 6.0.0
 * @internal
 */
export interface EnforceSelectedValueOptions<
  Option extends AutocompleteOption,
> {
  value: Option | readonly Option[] | null;
  visible: boolean;
  container: HTMLElement | null;
  popupRef: RefObject<HTMLElement>;
  comboboxRef: RefObject<EditableHTMLElement>;
  getOptionLabel: AutocompleteGetOptionLabel<Option>;
  availableOptions: readonly Option[];
  prevAvailableOptions: NonNullMutableRef<readonly Option[] | null>;
}

/**
 * This enforces that if the user clicks away, touches somewhere else on the
 * page, tabs to the next element, or focuses another element programmatically
 * the autocomplete will set the `value` back to the previous selected value if
 * there was one. i.e.
 *
 * Case 1:
 * - User selects "Apple"
 * - User hits backspace twice. Input displays "App"
 * - User clicks somewhere else on the page
 * - Input now displays "Apple" again
 *
 * Case 2:
 * - User selects "Apple"
 * - User clears the input
 * - User types "app"
 * - User clicks somewhere else on the page
 * - Input now displays "" and the value is set to `null`
 *
 * NOTE: This mutates the {@link EnforceSelectedValueOptions.prevAvailableOptions}
 *
 * @since 6.0.0
 * @internal
 */
export function enforceSelectedValue<Option extends AutocompleteOption>(
  options: EnforceSelectedValueOptions<Option>
): void {
  const {
    value,
    visible,
    container,
    popupRef,
    comboboxRef,
    getOptionLabel,
    availableOptions,
    prevAvailableOptions,
  } = options;

  if (!container) {
    return;
  }

  window.requestAnimationFrame(() => {
    if (
      container.contains(document.activeElement) ||
      popupRef.current?.contains(document.activeElement)
    ) {
      return;
    }

    if (visible) {
      // this makes it so that the options are not filtered again while closing
      prevAvailableOptions.current = availableOptions;
    }

    let label = "";
    if (typeof value === "string") {
      label = value;
    } else if (typeof value === "object" && value && !("length" in value)) {
      label = getOptionLabel(value);
    }

    triggerManualChangeEvent(comboboxRef.current, label);
  });
}
