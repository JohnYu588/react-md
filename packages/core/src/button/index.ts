export * from "./Button";
export * from "./ButtonUnstyled";
export * from "./FloatingActionButton";
