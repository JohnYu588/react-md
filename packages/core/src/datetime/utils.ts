/**
 * Since time input steps are based on seconds, this is a simple helper to
 * create the step in a human-readable form.
 *
 * @since 6.3.0
 */
export interface TimeFieldStepOptions {
  /**
   * @defaultValue `0`
   */
  seconds?: number;

  /**
   * @defaultValue `0`
   */
  minutes?: number;

  /**
   * @defaultValue `0`
   */
  hours?: number;
}

/**
 * Since time input steps are based on seconds, this is a simple helper to
 * create the step in a human-readable form.
 *
 * @example Simple Example
 * ```tsx
 * const step1 = getTimeStep({ minutes: 15 });
 * const step2 = getTimeStep({ hours: 1 });
 * const step3 = getTimeStep({ seconds: 15, minutes: 30, hours: 2 });
 * ```
 *
 * @since 6.3.0
 */
export function getTimeStep(
  step: TimeFieldStepOptions | "any" | number | undefined
): number | "any" | undefined {
  if (!step || typeof step === "string" || typeof step === "number") {
    return step;
  }

  const { hours = 0, minutes = 0, seconds = 0 } = step;

  const total = Math.abs(Math.round(seconds + minutes * 60 + hours * 60 * 60));
  return total === 0 ? undefined : total;
}
