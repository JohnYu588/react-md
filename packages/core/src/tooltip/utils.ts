import type { PositionAnchor, SimplePosition } from "../positioning";
import {
  ABOVE_CENTER_ANCHOR,
  BELOW_CENTER_ANCHOR,
  CENTER_LEFT_ANCHOR,
  CENTER_RIGHT_ANCHOR,
} from "../positioning";

/** @internal */
export function getAnchor(position: SimplePosition): PositionAnchor {
  switch (position) {
    case "above":
      return ABOVE_CENTER_ANCHOR;
    case "below":
      return BELOW_CENTER_ANCHOR;
    case "left":
      return CENTER_LEFT_ANCHOR;
    case "right":
      return CENTER_RIGHT_ANCHOR;
    default:
      throw new Error(`Invalid position: ${position}`);
  }
}
