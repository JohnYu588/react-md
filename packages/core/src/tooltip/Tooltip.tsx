"use client";
import { cnb } from "cnbuilder";
import type { HTMLAttributes } from "react";
import { forwardRef } from "react";
import { Portal } from "../portal/Portal";
import type { SimplePosition } from "../positioning/types";
import type {
  CSSTransitionComponentProps,
  TransitionActions,
} from "../transition/types";
import { useCSSTransition } from "../transition/useCSSTransition";
import { useEnsuredId } from "../useEnsuredId";
import { bem } from "../utils/bem";
import {
  DEFAULT_TOOLTIP_CLASSNAMES,
  DEFAULT_TOOLTIP_POSITION,
  DEFAULT_TOOLTIP_TIMEOUT,
} from "./constants";

const styles = bem("rmd-tooltip");

/**
 * @remarks \@since 6.0.0
 */
export interface TooltipClassNameOptions {
  className?: string;
  dense?: boolean;
  position: SimplePosition;
  disableLineWrap?: boolean;
}

/**
 * @remarks \@since 6.0.0
 */
export function tooltip(options: TooltipClassNameOptions): string {
  const { dense, position, className, disableLineWrap } = options;

  return cnb(
    styles({
      dense,
      [position]: true,
      nowrap: disableLineWrap,
    }),
    className
  );
}

/**
 * The base props for the `Tooltip` component. This can be extended when
 * creating custom tooltip implementations.
 *
 * @remarks
 * \@since 2.8.0 Supports the `RenderConditionalPortalProps`
 * \@since 6.0.0 No longer supports the `RenderConditionalPortalProps`.
 */
export interface TooltipProps
  extends HTMLAttributes<HTMLSpanElement>,
    CSSTransitionComponentProps,
    TransitionActions {
  visible: boolean;

  /**
   * @defaultValue `false`
   */
  dense?: boolean;

  /**
   * @defaultValue `DEFAULT_TOOLTIP_POSITION`
   * @see {@link DEFAULT_TOOLTIP_POSITION}
   */
  position?: SimplePosition;

  /**
   * @defaultValue `false`
   */
  disablePortal?: boolean;

  /**
   * Set this to `true` to add `whitespace: nowrap` to the tooltip.
   *
   * This prop is useful for tooltips that are positioned near the edge of the
   * viewport with `"above"` or `"below"` so that the tooltip no longer aligns
   * the center of the tooltip with the center of the tooltipped element.
   *
   * @defaultValue `false`
   */
  disableLineWrap?: boolean;
}

/**
 * **Client Component**
 *
 * This is the base tooltip component that can only be used to render a tooltip
 * with an animation when the visibility changes. If this component is used, you
 * will need to manually add all the event listeners and triggers to change the
 * `visible` prop.
 *
 * @example
 * Simple Usage
 * ```tsx
 * import { Button, useTooltip, Tooltip } from "@react-md/core";
 *
 * function Example() {
 *   const { elementProps, tooltipProps } = useTooltip();
 *
 *   return (
 *     <>
 *       <Button {...elementProps}>Button</Button>
 *       <Tooltip {...tooltipProps}>
 *         Tooltip Content
 *       </Tooltip>
 *     </>
 *   );
 * }
 * ```
 */
export const Tooltip = forwardRef<HTMLSpanElement, TooltipProps>(
  function Tooltip(props, nodeRef) {
    const {
      id: propId,
      dense,
      visible,
      children,
      appear,
      enter,
      exit,
      onEnter,
      onEntering,
      onEntered,
      onExit,
      onExiting,
      onExited,
      timeout = DEFAULT_TOOLTIP_TIMEOUT,
      classNames = DEFAULT_TOOLTIP_CLASSNAMES,
      className,
      position = DEFAULT_TOOLTIP_POSITION,
      temporary,
      exitedHidden = !temporary,
      disableLineWrap,
      disablePortal: propDisablePortal,
      ...remaining
    } = props;
    const id = useEnsuredId(propId, "tooltip");

    const { rendered, elementProps, disablePortal } = useCSSTransition({
      nodeRef,
      appear,
      enter,
      exit,
      transitionIn: visible,
      timeout,
      classNames,
      className: tooltip({
        dense,
        position,
        className,
        disableLineWrap,
      }),
      onEnter,
      onEntering,
      onEntered,
      onExit,
      onExiting,
      onExited,
      temporary,
      exitedHidden,
      disablePortal: propDisablePortal,
    });

    return (
      <Portal disabled={disablePortal}>
        {rendered && (
          <span {...remaining} {...elementProps} id={id} role="tooltip">
            {children}
          </span>
        )}
      </Portal>
    );
  }
);
