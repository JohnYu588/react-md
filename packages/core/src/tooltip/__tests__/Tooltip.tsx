import {
  act,
  fireEvent,
  render as baseRender,
  waitFor,
} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import type { ReactElement } from "react";
import { Button } from "../../button";
import { CoreProviders } from "../../CoreProviders";
import { FontIcon } from "../../icon";
import { parseCssLengthUnit } from "../../utils";
import {
  DEFAULT_TOOLTIP_DENSE_SPACING,
  DEFAULT_TOOLTIP_SPACING,
} from "../constants";
import type { TooltipProps } from "../Tooltip";
import { Tooltip } from "../Tooltip";
import type { TooltipHookOptions } from "../useTooltip";
import { useTooltip } from "../useTooltip";

const render = (ui: ReactElement): ReturnType<typeof baseRender> =>
  baseRender(ui, {
    wrapper: ({ children }) => (
      <CoreProviders elementInteractionMode="none">{children}</CoreProviders>
    ),
  });

interface TestProps extends TooltipHookOptions {
  tooltip?: Partial<TooltipProps>;
}

function Test(props: TestProps): ReactElement {
  const { tooltip, ...options } = props;
  const {
    elementProps,
    tooltipProps,
    animatedOnce,
    hideTooltip,
    setVisible,
    visible,
  } = useTooltip(options);

  return (
    <>
      <Button {...elementProps}>Button</Button>
      <Tooltip {...tooltipProps} {...tooltip}>
        Tooltip
      </Tooltip>
      <Button onClick={hideTooltip}>Hide Tooltip</Button>
      <Button onClick={() => setVisible((prev) => !prev)}>
        Toggle Tooltip
      </Button>
      <div data-testid="animatedOnce">{`${animatedOnce}`}</div>
      <div data-testid="visible">
        {visible ? <FontIcon>favorite</FontIcon> : <FontIcon>close</FontIcon>}
      </div>
    </>
  );
}

describe("Tooltip", () => {
  afterEach(() => {
    jest.clearAllTimers();
    jest.useRealTimers();
  });

  it("should display the tooltip after hovering for 1s by default", () => {
    jest.useFakeTimers();
    const { getByRole } = render(<Test />);
    const button = getByRole("button", { name: "Button" });
    const tooltip = getByRole("tooltip", { hidden: true });
    expect(tooltip).toBeInTheDocument();

    fireEvent.mouseEnter(button);
    expect(tooltip).toHaveAttribute("hidden");

    act(() => {
      jest.advanceTimersByTime(800);
    });
    expect(tooltip).toHaveAttribute("hidden");

    fireEvent.mouseLeave(button);
    act(() => {
      jest.runAllTimers();
    });
    expect(tooltip).toHaveAttribute("hidden");

    fireEvent.mouseEnter(button);
    act(() => {
      jest.runAllTimers();
    });
    expect(tooltip).not.toHaveAttribute("hidden");
  });

  it("should display the tooltip after focusing for 1s by default only in keyboard mode", () => {
    jest.useFakeTimers();
    const { getByRole } = render(<Test />);
    const button = getByRole("button", { name: "Button" });
    const tooltip = getByRole("tooltip", { hidden: true });

    fireEvent.focus(button);
    act(() => {
      jest.advanceTimersByTime(1000);
    });
    expect(tooltip).toHaveAttribute("hidden");
    act(() => {
      jest.runAllTimers();
    });
    expect(tooltip).toHaveAttribute("hidden");

    fireEvent.keyDown(document.body);
    fireEvent.focus(button);
    act(() => {
      jest.advanceTimersByTime(1000);
    });
    expect(tooltip).not.toHaveAttribute("hidden");
    act(() => {
      jest.runAllTimers();
    });
    expect(tooltip).toMatchSnapshot();

    fireEvent.blur(button);
    act(() => {
      jest.runOnlyPendingTimers();
    });
    expect(tooltip).toHaveClass("rmd-tooltip--exit");

    act(() => {
      jest.runAllTimers();
    });
    expect(tooltip).toHaveAttribute("hidden");
    expect(tooltip).toMatchSnapshot();
  });

  it("should show the tooltip after touching for the hoverTime or if the onContextMenu is called on mobile", () => {
    // some mobile devices will trigger the context menu for long presses, so
    // need to make sure that does not happen for tooltipped elements on mobile
    jest.useFakeTimers();

    const { getByRole } = render(<Test />);
    const button = getByRole("button", { name: "Button" });
    const tooltip = getByRole("tooltip", { hidden: true });

    fireEvent.touchStart(button);
    expect(tooltip).toHaveAttribute("hidden");
    act(() => {
      jest.advanceTimersByTime(1000);
    });
    expect(tooltip).not.toHaveAttribute("hidden");
    expect(tooltip).toHaveClass("rmd-tooltip--enter");

    act(() => {
      jest.runAllTimers();
    });
    expect(tooltip).not.toHaveAttribute("hidden");
    expect(tooltip).not.toHaveClass("rmd-tooltip--enter");

    fireEvent.touchEnd(button);
    expect(tooltip).toHaveClass("rmd-tooltip--exit");

    act(() => {
      jest.runAllTimers();
    });
    expect(tooltip).toHaveAttribute("hidden");

    fireEvent.click(document.body);
    fireEvent.mouseEnter(button);
    fireEvent.contextMenu(button);
    act(() => {
      jest.runAllTimers();
    });
    expect(tooltip).toHaveAttribute("hidden");

    fireEvent.touchStart(button);
    expect(tooltip).toHaveAttribute("hidden");
    act(() => {
      jest.advanceTimersByTime(800);
    });
    expect(tooltip).toHaveAttribute("hidden");

    const empty = jest.spyOn(Selection.prototype, "empty");
    fireEvent.contextMenu(button);
    act(() => {
      jest.runAllTimers();
    });
    expect(tooltip).not.toHaveAttribute("hidden");
    expect(empty).not.toHaveBeenCalled();

    const empty2 = jest.fn();
    jest.spyOn(window, "getSelection").mockReturnValue({
      empty: empty2,
      anchorNode: {
        parentElement: button,
      },
    } as unknown as Selection);

    fireEvent.contextMenu(button);
    expect(empty2).toHaveBeenCalled();
  });

  it("should support controlling the visibility manually with the setVisible and hideTooltip returned functions", async () => {
    const user = userEvent.setup();
    const { getByRole } = render(<Test />);

    const button = getByRole("button", { name: "Button" });
    const tooltip = getByRole("tooltip", { hidden: true });
    const toggleTooltip = getByRole("button", { name: "Toggle Tooltip" });
    expect(tooltip).toBeInTheDocument();
    expect(button).not.toHaveAttribute("aria-describedby");

    await user.click(toggleTooltip);
    expect(tooltip).not.toHaveAttribute("hidden");
    expect(button).toHaveAttribute("aria-describedby");
    expect(tooltip).toMatchSnapshot();
    await waitFor(() => {
      expect(tooltip).not.toHaveClass("rmd-tooltip--enter");
    });
    expect(tooltip).toMatchSnapshot();

    await user.click(toggleTooltip);
    await waitFor(() => {
      expect(tooltip).toHaveAttribute("hidden");
    });
    expect(tooltip).toMatchSnapshot();

    await user.click(toggleTooltip);
    await waitFor(() => {
      expect(tooltip).not.toHaveClass("rmd-tooltip--enter");
    });

    await user.click(getByRole("button", { name: "Hide Tooltip" }));
    await waitFor(() => {
      expect(tooltip).toHaveAttribute("hidden");
    });
    expect(tooltip).toMatchSnapshot();
  });

  it("should support rendering the tooltip temporarily instead of using hidden", async () => {
    const user = userEvent.setup();
    const { getByRole } = render(
      <Test tooltip={{ temporary: true }} hoverTime={0} />
    );

    expect(() => getByRole("tooltip", { hidden: true })).toThrow();

    await user.tab();
    await waitFor(() => {
      expect(() => getByRole("tooltip", { name: "Tooltip" })).not.toThrow();
    });
  });

  it("should close the tooltip whenever the escape key is pressed", async () => {
    const user = userEvent.setup();
    const { getByRole } = render(<Test hoverTime={0} />);
    const tooltip = getByRole("tooltip", { hidden: true });

    await user.tab();
    await waitFor(() => {
      expect(tooltip).not.toHaveAttribute("hidden");
      expect(tooltip).not.toHaveClass("rmd-tooltip--enter");
    });

    await user.keyboard("[Escape]");
    await waitFor(() => {
      expect(tooltip).toHaveAttribute("hidden");
    });
  });

  it("should close the tooltip whenever the page is scrolled", async () => {
    const user = userEvent.setup();
    const { getByRole } = render(<Test hoverTime={0} />);
    const tooltip = getByRole("tooltip", { hidden: true });

    await user.tab();
    await waitFor(() => {
      expect(tooltip).not.toHaveAttribute("hidden");
      expect(tooltip).not.toHaveClass("rmd-tooltip--enter");
    });

    fireEvent.scroll(document.body);
    await waitFor(() => {
      expect(tooltip).toHaveAttribute("hidden");
    });
  });

  it("should hide the tooltip when the page becomes inactive or blurs and not show the tooltip again on focus", async () => {
    const user = userEvent.setup();
    const { getByRole } = render(<Test hoverTime={0} />);
    const button = getByRole("button", { name: "Button" });
    const tooltip = getByRole("tooltip", { hidden: true });

    await user.tab();
    expect(tooltip).not.toHaveAttribute("hidden");
    await waitFor(() => {
      expect(tooltip).not.toHaveClass("rmd-tooltip--enter");
    });

    fireEvent.blur(window);
    await waitFor(() => {
      expect(tooltip).toHaveAttribute("hidden");
    });

    fireEvent.focus(button);
    expect(tooltip).toHaveAttribute("hidden");

    await user.tab();
    await user.tab({ shift: true });
    expect(document.activeElement).toBe(button);
    expect(tooltip).not.toHaveAttribute("hidden");
    await waitFor(() => {
      expect(tooltip).not.toHaveClass("rmd-tooltip--enter");
    });

    jest
      .spyOn(document, "visibilityState", "get")
      .mockReturnValueOnce("hidden")
      .mockReturnValueOnce("visible");

    act(() => {
      document.dispatchEvent(new Event("visibilitychange"));
    });
    await waitFor(() => {
      expect(tooltip).toHaveAttribute("hidden");
    });

    act(() => {
      document.dispatchEvent(new Event("visibilitychange"));
    });
    fireEvent.focus(button);
    expect(tooltip).toHaveAttribute("hidden");
  });

  it("should use the position prop when provided instead of determining a position within the viewport", async () => {
    const vh = 768;
    expect(window.innerHeight).toBe(vh);

    const user = userEvent.setup();
    const { getByRole } = render(<Test hoverTime={0} position="above" />);
    const button = getByRole("button", { name: "Button" });
    const tooltip = getByRole("tooltip", { hidden: true });

    const baseRect = document.body.getBoundingClientRect();
    jest.spyOn(button, "getBoundingClientRect").mockReturnValue({
      ...baseRect,
      top: vh / 2,
    });

    await user.tab();
    expect(tooltip).not.toHaveAttribute("hidden");
    expect(tooltip).toHaveClass("rmd-tooltip--above");

    await user.tab();
    await waitFor(() => {
      expect(tooltip).toHaveAttribute("hidden");
    });

    jest.spyOn(button, "getBoundingClientRect").mockReturnValue({
      ...baseRect,
      top: 0,
    });
    await user.tab({ shift: true });
    expect(tooltip).not.toHaveAttribute("hidden");
    expect(tooltip).toHaveClass("rmd-tooltip--above");
  });

  it("should never display the tooltip if the disabled option is true", async () => {
    const user = userEvent.setup();
    const { getByRole } = render(<Test hoverTime={0} disabled />);
    const button = getByRole("button", { name: "Button" });
    const tooltip = getByRole("tooltip", { hidden: true });

    await user.tab();
    expect(tooltip).toHaveAttribute("hidden");

    await user.tab();
    expect(tooltip).toHaveAttribute("hidden");

    await user.hover(button);
    expect(tooltip).toHaveAttribute("hidden");

    await user.unhover(button);
    expect(tooltip).toHaveAttribute("hidden");

    fireEvent.touchStart(button);
    expect(tooltip).toHaveAttribute("hidden");

    fireEvent.touchEnd(button);
    expect(tooltip).toHaveAttribute("hidden");
  });

  it("should automatically attempt to determine the spacing based on the computed style of the spacing custom property", async () => {
    const user = userEvent.setup();
    const { getByRole, rerender } = render(<Test hoverTime={0} />);
    const button = getByRole("button", { name: "Button" });
    const tooltip = getByRole("tooltip", { hidden: true });

    const spacing = "0.825rem";
    const spacingPixels = parseCssLengthUnit({ value: spacing });

    const getComputedStyle = jest
      .spyOn(window, "getComputedStyle")
      .mockImplementation(
        // @ts-expect-error
        () => ({
          getPropertyValue() {
            return spacing;
          },
        })
      );

    await user.hover(button);
    await waitFor(() => {
      expect(tooltip).not.toHaveAttribute("hidden");
      expect(tooltip).not.toHaveClass("rmd-tooltip--enter");
    });
    expect(tooltip.style.top).toBe(`${spacingPixels}px`);

    await user.unhover(button);
    await waitFor(() => {
      expect(tooltip).toHaveAttribute("hidden");
    });

    getComputedStyle.mockImplementation(
      // @ts-expect-error
      () => ({
        getPropertyValue() {
          return null;
        },
      })
    );

    await user.hover(button);
    await waitFor(() => {
      expect(tooltip).not.toHaveAttribute("hidden");
      expect(tooltip).not.toHaveClass("rmd-tooltip--enter");
    });

    const defaultSpacingPixels = parseCssLengthUnit({
      value: DEFAULT_TOOLTIP_SPACING,
    });
    expect(tooltip.style.top).toBe(`${defaultSpacingPixels}px`);

    await user.unhover(button);
    await waitFor(() => {
      expect(tooltip).toHaveAttribute("hidden");
    });

    rerender(<Test dense disableAutoSpacing hoverTime={0} />);
    expect(tooltip).toBeInTheDocument();
    await user.hover(button);
    await waitFor(() => {
      expect(tooltip).not.toHaveAttribute("hidden");
      expect(tooltip).not.toHaveClass("rmd-tooltip--enter");
    });

    const denseSpacingPixels = parseCssLengthUnit({
      value: DEFAULT_TOOLTIP_DENSE_SPACING,
    });
    expect(tooltip.style.top).toBe(`${denseSpacingPixels}px`);
  });
});
