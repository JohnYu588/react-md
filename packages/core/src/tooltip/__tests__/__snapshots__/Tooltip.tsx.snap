// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip should display the tooltip after focusing for 1s by default only in keyboard mode 1`] = `
<span
  class="rmd-tooltip rmd-tooltip--below rmd-tooltip--visible"
  id=":r2:-tooltip"
  role="tooltip"
  style="position: fixed; left: 16px; top: 24px;"
>
  Tooltip
</span>
`;

exports[`Tooltip should display the tooltip after focusing for 1s by default only in keyboard mode 2`] = `
<span
  class="rmd-tooltip rmd-tooltip--below rmd-display-none"
  id=":r2:-tooltip"
  role="tooltip"
  style="position: fixed; left: 16px; top: 24px;"
>
  Tooltip
</span>
`;

exports[`Tooltip should support controlling the visibility manually with the setVisible and hideTooltip returned functions 1`] = `
<span
  class="rmd-tooltip rmd-tooltip--below rmd-tooltip--enter rmd-tooltip--visible"
  id=":r6:-tooltip"
  role="tooltip"
  style="position: fixed; left: 16px; top: 24px;"
>
  Tooltip
</span>
`;

exports[`Tooltip should support controlling the visibility manually with the setVisible and hideTooltip returned functions 2`] = `
<span
  class="rmd-tooltip rmd-tooltip--below rmd-tooltip--visible"
  id=":r6:-tooltip"
  role="tooltip"
  style="position: fixed; left: 16px; top: 24px;"
>
  Tooltip
</span>
`;

exports[`Tooltip should support controlling the visibility manually with the setVisible and hideTooltip returned functions 3`] = `
<span
  class="rmd-tooltip rmd-tooltip--below rmd-display-none"
  id=":r6:-tooltip"
  role="tooltip"
  style="position: fixed; left: 16px; top: 24px;"
>
  Tooltip
</span>
`;

exports[`Tooltip should support controlling the visibility manually with the setVisible and hideTooltip returned functions 4`] = `
<span
  class="rmd-tooltip rmd-tooltip--below rmd-display-none"
  id=":r6:-tooltip"
  role="tooltip"
  style="position: fixed; left: 16px; top: 24px;"
>
  Tooltip
</span>
`;
